### PR TITLE
To-Do Lists and Prisma Refactor

### PR DESCRIPTION
**Pull request by LINK1733**
Switched from using Mongo to Prisma due to needing the relational db more than Mongo.

Added functionality with IGDB and they advise to run a local copy of their db and use webhooks to get updates to it. Haven't setup the update part but that will come later. For now you will run the file for creating the local copy of the igdb database if changes are needed.

Added todo lists that are assigned to the user that created the list. This will also only show to that user if they are signed in. You can create lists and then add games to the lists by searching for a game and selecting the desired game in the drop down. Styling needs to be fine tuned still. Can click on the x to delete the item/list if needed.

Cleaned up the npm packages to accurately reflect was is being used.

### CODE REVIEW

File: src/static/main.js [hunk 1]
```diff
@@ -1,49 +1,348 @@
...
+	};
+
+	// const handleToggle = (id) => {
+	// 	let mapped = gameList.map((gameName) => {
+	// 		return gameName.id == id
+	// 			? { ...gameName, complete: !gameName.complete }
+	// 			: { ...gameName };
+	// 	});
+	// 	setGameList(mapped);
+	// };
```
Comment (LINK1733) [inline src/static/main.js:327]: This is hidden since I'm not sure on if I want to keep the completed option in the way that I currently had implemented it. Since each game will have it's own page in the end the toggle to do a strike-through probably won't be needed anymore.

File: src/static/Main.scss [hunk 2]
```diff
@@ -0,0 +1,7 @@
+.strike {
+	text-decoration: line-through;
+}
```
Comment (LINK1733) [inline src/static/Main.scss:3]: If I end up not doing the toggle to complete this won't be needed anymore.

File: prisma/schema.prisma [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+model Game {
+  id           String   @id @default(cuid())
+  complete     Boolean
+  gameName     String
+  author       User     @relation(fields: [authorId], references: [id])
+  authorId     String
+  parentList   GameList @relation(fields: [parentListId], references: [id])
+  parentListId String
+  cover        String?
```
Comment (sheodox) [inline prisma/schema.prisma:40]: I'd recommend not including the game name/cover stuff on here, that's on the `igdbGame`, it'd be better to just have a reference to that table. Use the relational strengths of a relational DB! (This means if a game is updated via webhook you'll get updates to all the games in this table without having to subsequently update the Game models (there's also no reference back to the igdbGame from here so you couldn't really update these anyway).
```
  igdbGameId String
  igdbGame igdbGame @relation(fields: [igdbGameId], references: [id])
```

Then when selecting the games list you could do:
```
	const gameLists = await prisma.gameList.findMany({
		where: {
			authorId: req.user.id,
		},
		include: {
			games: {
				include: {
					igdbGame: true
				}
			}
		},
	});
```

Then get the game data in react by `game.igdbGame.cover ` or whatever.

File: src/server/controllers/gameList.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+				},
+				parentList: {
+					connect: {
+						id: req.body.parentList,
+					},
+				},
+				cover: coverPhoto,
+				complete: false,
+			},
+		});
```
Comment (sheodox) [inline src/server/controllers/gameList.js:65]: Like I mentioned in another comment, I would recommend not duplicating the `igdbGame` data, keep it relational! Also if I were writing this I'd only accept the `gameId` from `req.body`, and query for the cover/gameName, not just the cover, so the user can't lie and rename the game by messing with the http request lol. Again not really relevant if you switch to just referencing the igdbGame table.

File: src/server/controllers/igdbWebhook.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	).request('/games/count');
+
+	const total = totalGames.data.count;
+	let data = [];
+	let games = [];
+	try {
+		for (i = 0; i < total; i += 500) {
+			const newGames = await igdb(
+				`${process.env.TWITCH_CLIENT_ID}`,
+				`${process.env.TWITCH_APP_ACCESS_TOKEN}`
```
Comment (sheodox) [inline src/server/controllers/igdbWebhook.js:22]: These are just strings, there's no point in putting them in a template string when you're not putting anything else in the template string. I'm seeing you do this all over, why? I'm confused, this isn't doing anything.

File: src/server/controllers/igdbWebhook.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+		`${process.env.TWITCH_CLIENT_ID}`,
+		`${process.env.TWITCH_APP_ACCESS_TOKEN}`
+	).request('/games/count');
+
+	const total = totalGames.data.count;
+	let data = [];
+	let games = [];
+	try {
+		for (i = 0; i < total; i += 500) {
+			const newGames = await igdb(
```
Comment (sheodox) [inline src/server/controllers/igdbWebhook.js:20]: Aren't you rate limited by igdb to like 4 requests per second? Does this library handle that?

File: src/static/main.js [hunk 7]
```diff
@@ -1,49 +1,348 @@
...
+													'thumb',
+													'micro'
+												)}
+												onClick={handleSubmit(
+													searchResult
+												)}
+												id={searchResult.id}
+											/>
+										)}
+										<label
```
Comment (sheodox) [inline src/static/main.js:215]: This GameList component is getting really long and really indented. I'd recommend splitting things up so it's easier to work on. Giant components are a headache to maintain because it's hard to understand the full picture to make changes, but if it's broken down into smaller pieces it's easier to understand the whole piece you need to work on.

File: src/static/main.js [hunk 8]
```diff
@@ -1,49 +1,348 @@
...
+		</Router>
+	);
 }
-// ms-auto mt-2 btn btn-primary btn-small
-ReactDom.render(
-        <App />,
-    document.querySelector('#app-root')
-)
\ No newline at end of file
+ReactDom.render(<App />, document.querySelector('#app-root'));
```
Comment (sheodox) [inline src/static/main.js]: I'd recommend splitting up `main.js` into a bunch of individual files. Generally one file per component (except if you need something small and you don't think it deserves its own file), and keep them organized in logical folders. Having everything in one file gets to be hard to work on. In my `main.js` files I generally only have css imports and initializing the app (like your `ReactDom.render` line here.) then everything else is in their own files.

You can poke around [Overseer's components](https://github.com/sheodox/Overseer/tree/master/src/static/components) to see what sort of stuff I generally break out into separate components/files.

File: src/static/Main.scss (hunk 2 above)
Comment (LINK1733) [reply src/static/Main.scss:3]: it was when i first did the todo lists. i made a checkbox and if you clicked the checkbox it would strike through the game to show it as completed.

File: package.json [hunk 9]
```diff
@@ -1,45 +1,51 @@
...
-  }
+	"name": "game-tracker",
+	"version": "0.0.1",
+	"description": "Site for making lists of games to play and things to do in them.",
+	"main": "app.js",
+	"scripts": {
+		"build-static:dev": "webpack --mode development",
+		"build-static:prod": "webpack --mode production",
+		"Start App": "nodemon src/server/app.js --watch src/server",
+		"Prisma Studio": "npx prisma studio"
```
[the remark this answers is not in the document]
Comment (LINK1733) [reply package.json:10]: ah gotcha. it's been changed.

File: .gitignore [hunk 10]
```diff
@@ -1,5 +1,4 @@
 node_modules
-.vscode
+# Keep environment variables out of version control
 .env
-.idea
```
[the remark this answers is not in the document]
Comment (LINK1733) [reply .gitignore:4]: not sure honestly. i never really look in that file so idk when that would have been taken out even. i'll go i don't have any .vscode files and i'm assuming .idea was from the ide you used to use correct?

File: package.json [hunk 11]
```diff
@@ -1,45 +1,51 @@
...
+		"joi": "^17.4.0",
+		"passport": "^0.4.1",
+		"passport-local": "^1.0.0",
+		"react": "^17.0.2",
+		"react-dom": "^17.0.2",
+		"react-router-dom": "^5.2.0",
+		"sass": "^1.35.1",
+		"sass-loader": "^12.1.0",
+		"style-loader": "^2.0.0",
+		"util": "^0.12.4",
```
[the remark this answers is not in the document]
Comment (LINK1733) [reply package.json:42]: i don't think i put it in the webpack. so idk if it was something you put in there when you set the webpack stuff up?

File: src/static/main.js (hunk 8 above)
Comment (LINK1733) [reply src/static/main.js]: good news! i broke up the file in the stuff i've been working on so that'll show up on my next push i do for the stuff i've been working on.

File: src/static/main.js (hunk 7 above)
Comment (LINK1733) [reply src/static/main.js:215]: I'll have to see how i can break it up more. i might be able to take some of the functions and make them their own files?

File: src/server/controllers/igdbWebhook.js (hunk 6 above)
Comment (LINK1733) [reply src/server/controllers/igdbWebhook.js:20]: it should be but it doesn't stop this from running. its just not the fastest but thats probably because its stopping me cause im requesting too much? i'd have to do a set timeout to like 250ms right?

File: src/server/controllers/igdbWebhook.js (hunk 5 above)
Comment (LINK1733) [reply src/server/controllers/igdbWebhook.js:22]: for some reason i didn't think it worked if i didn't put them in that. I'll try it without them though.

Comment (LINK1733) [reply src/server/controllers/igdbWebhook.js:22]: It might also be from when I was using axios and had to put Bearer in front of the access token.

File: prisma/schema.prisma (hunk 3 above)
Comment (LINK1733) [reply prisma/schema.prisma:40]: good point. i'll have to move stuff around and change things but i should be able to get this switched over.

File: src/server/controllers/gameList.js (hunk 4 above)
Comment (LINK1733) [reply src/server/controllers/gameList.js:65]: like noted in the other comment i'll work on getting the table's switched around and accommodate accordingly

File: .gitignore (hunk 10 above)
[the remark this answers is not in the document]
Comment (sheodox) [reply .gitignore:4]: Yeah `.idea` is webstorm and the other Jetbrains IDEs. I'm using neovim now, but I still have that folder, best to keep it around in the `.gitignore`

File: package.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (sheodox) [reply package.json:42]: I didn't do it, it's in your PR. This is what I was referring to, in your `webpack.config.js`

image

File: src/server/controllers/igdbWebhook.js (hunk 5 above)
Comment (sheodox) [reply src/server/controllers/igdbWebhook.js:22]: Gotcha, you shouldn't need to do it this way, unless you wanted something like `undefined` to turn into the string `"undefined"` lol.

File: src/server/controllers/igdbWebhook.js (hunk 6 above)
Comment (sheodox) [reply src/server/controllers/igdbWebhook.js:20]: The library is probably handling the rate limiting then if you're not seeing any errors.

If it wasn't handling the rate limiting then yes you'd need a timeout. You can make it nice with promises like:
```js
await new Promise(resolve => setTimeout(resolve, 250));
```

File: src/static/main.js (hunk 7 above)
Comment (sheodox) [reply src/static/main.js:215]: If I were you'd I'd just move the search stuff into its own component. It takes up like half of this component by itself.

Also, I just looked at the actual markup here. The images and stuff _look_ good, but it's not [semantic HTML](https://developer.mozilla.org/en-US/docs/Glossary/semantics#semantics_in_html). If you have a click handler on something you really want it to be an actual clickable element. Again if you think about screen readers there's no way for a screen reader to know you can click on these just because it's an element with a click handler, the HTML has to mean something. So instead of binding click handlers to a label and an img you should put it in an actual `<button>`. I don't know if you can just use a label on an image anyway? [Doesn't look like it actually](https://html.spec.whatwg.org/multipage/forms.html#category-label):

> Some elements, not all of them form-associated, are categorized as labelable elements. These are elements that can be associated with a label element. button, input (if the type attribute is not in the Hidden state), meter, output, progress, select, textarea, form-associated custom elements

Basically you should only use labels with input fields of some sort.

File: .gitignore (hunk 10 above)
[the remark this answers is not in the document]
Comment (LINK1733) [reply .gitignore:4]: sounds good i'll throw them back in there

File: prisma/schema.prisma (hunk 3 above)
Comment (LINK1733) [reply prisma/schema.prisma:40]: Got this switched over now.

File: src/static/Main.scss (hunk 2 above)
Comment (LINK1733) [reply src/static/Main.scss:3]: I ended up removing the toggle stuff so I removed this from the scss as well.

File: src/server/controllers/igdbWebhook.js (hunk 5 above)
Comment (LINK1733) [reply src/server/controllers/igdbWebhook.js:22]: removed the quotes and it works fine as was expected.

File: src/static/main.js (hunk 7 above)
Comment (LINK1733) [reply src/static/main.js:215]: Gotcha. I'll get it switched over to a button.

File: src/server/controllers/igdbWebhook.js (hunk 6 above)
Comment (LINK1733) [reply src/server/controllers/igdbWebhook.js:20]: so would I put the promise as part the igdb chain?

File: src/static/main.js (hunk 7 above)
Comment (LINK1733) [reply src/static/main.js]: Ok this should be switched over to a button now but is also in it's own file in the components folder.

File: src/server/controllers/igdbWebhook.js (hunk 6 above)
Comment (LINK1733) [reply src/server/controllers/igdbWebhook.js:23]: Gotcha that makes sense. Yea i'll just leave it as is then.

File: prisma/schema.prisma [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+model igdbGame {
+  id            Int     @id @unique
+  name          String
+  cover         String?
+  versionParent String?
+  parentGame    String?
+}
```
[the remark this answers is not in the document]
Comment (LINK1733) [reply prisma/schema.prisma]: I haven't gotten around to indexing yet. Kinda forgot about it but I'll look into it still.

File: src/static/Main.scss (hunk 2 above)
Comment (LINK1733) [reply src/static/Main.scss]: This update should be on here now.

File: src/static/main.js (hunk 1 above)
Comment (LINK1733) [reply src/static/main.js]: This has been removed and updated on here now.

**Comment by LINK1733** [comment]
> 
> 
> You said you'd leave it as-is but then added this anyway!

did i forget to remove something?

**Comment by LINK1733** [comment]
I sure did forget to remove it hah. i removed it now.

File: prisma/schema.prisma (hunk 12 above)
[the remark this answers is not in the document]
Comment (LINK1733) [reply prisma/schema.prisma]: Added indexing. So I believe that should be the last of the PR stuff?

File: src/static/main.js (hunk 7 above)
Comment (LINK1733) [reply src/static/main.js]: Broke this file up to be its own component.

File: src/static/components/gameSearch.js [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+	const [searchResult, setSearchResult] = useState([]);
+
+	useEffect(() => {
+		const timeoutId = setTimeout(() => {
+			const searchInput = {
+				searchInput: gameForm,
+			};
+			if (gameForm.length === 0) {
+				setSearchResult(0);
```
Comment (sheodox) [inline src/static/components/gameSearch.js:15]: It looks like you're otherwise setting this variable to an array in the 'success case', I'd recommend staying away from changing the variable type. If you want an empty result set I'd do `setSearchResult([])` and check the length instead of checking if it's `0` in your render code. A `0` by itself isn't super descriptive and I think that sort of thing should be avoided!

Comment (LINK1733) [reply src/static/components/gameSearch.js:15]: oh yea idk why i set it to that. it should probably just be an empty string instead of 0. i probably got mixed up with the length thing.

Comment (LINK1733) [reply src/static/components/gameSearch.js:15]: Or I could yea just do [] like you said lol.

Comment (LINK1733) [reply src/static/components/gameSearch.js]: Ok this should be changed now.

File: src/static/components/gameSearch.js [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+				className="w-100 my-1 "
+				type="text"
+				id="new-game-input"
+				name="new-game-input"
+				onChange={handleChange}
+				placeholder="New Game"
+			/>
+			{searchResult != 0 && gameForm != '' && (
+				<div id="searchResults" className="shadow border bg-light p-3">
+					{searchResult.map((searchResult) => {
```
Comment (sheodox) [inline src/static/components/gameSearch.js:59]: You changed the empty set to `[]`, but your check here still ensures it's not explicitly `0`, not `!searchResult.length`. If it's empty then you'll still get this `div#searchResults` div where you wouldn't have before.

File: src/static/components/gameSearch.js [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+										id={searchResult.id}
+									/>
+								)}
+								<label
+									onClick={handleSubmit(searchResult)}
+									className="form-check-label flex-grow-1 px-2 my-auto"
+									htmlFor={searchResult.id}
+								>
+									{searchResult.name}
+								</label>
```
Comment (sheodox) [inline src/static/components/gameSearch.js:81]: I thought you said you had turned this into a button but it doesn't seem that change made it to the PR?

Additionally, I see your search results are inside the form, but these here are search results (basically the result of the form) so in my opinion I think they should be outside of the form element.

Comment (LINK1733) [reply src/static/components/gameSearch.js:81]: Oh I changed it for the game items but forgot to do it for the search results. so there were 2 places I needed to change it but didn't realize that. Whoops!

Comment (LINK1733) [reply src/static/components/gameSearch.js:81]: Ok I have the label updated to be a button in the list that shows up when you search for a game to add.

File: src/static/components/gameSearch.js (hunk 14 above)
Comment (LINK1733) [reply src/static/components/gameSearch.js:59]: so when I switch it to !searchResult.length I get a small white box that shows up when I type but when it actually searches for a game like Persona 3 there's no search results that show up. But if I try something like searchResult != [] it shows the little box cause there's no searchResults to show but this time it shows the searchResults when it's supposed to. So i'm not sure, I get that the check isn't right but it seems to work fine with how it is when it looks for != 0.

Comment (sheodox) [reply src/static/components/gameSearch.js:59]: Oh oops, I meant `searchResult.length`, not the `!searchResult.length`, had it backwards!

Also you can't check if an array is equal to another array with just an equality operator, as objects and arrays will only check if it's the same _instance_ of an object! 

image

In order to check if an array or object equals each other you need to check that every element in each matches the matching on in the other. Like for arrays you'd need something like this:
image

Comment (sheodox) [reply src/static/components/gameSearch.js:59]: Basically when you changed it to `searchResult != []` that's never going to fail, because by doing `[]` there you're making a new array, and you're asking if searchResult is the same array you _just made_, which would be impossible. Does that make sense?

Comment (LINK1733) [reply src/static/components/gameSearch.js:59]: uhhh 
image

Comment (LINK1733) [reply src/static/components/gameSearch.js:59]: figured it out we good.

Comment (sheodox) [reply src/static/components/gameSearch.js:59]: Try `!!searchResult.length`. I think react tries to render numbers, but not booleans or null. If the first half of your `x && y` is falsey the result of that expression is that falsey value. For example see this:
image

So if we're checking `searchResult.length && somethingElse` It's evaluating to `somethingElse` if `searchResult.length` is greater than 0, but `0` if it's not.

`!!something` is some shorthand for turning something into a boolean by just double negating it.

Comment (sheodox) [reply src/static/components/gameSearch.js:59]: So if you're double negating it you're casting it as a boolean, so you're handing react the `0` instead of the JSX you're giving it otherwise.

image

Comment (LINK1733) [reply src/static/components/gameSearch.js]: Oh didn't see this till now. Should I try this vs what I changed it to?

Comment (sheodox) [reply src/static/components/gameSearch.js]: It's basically the same, shouldn't matter!

Comment (LINK1733) [reply src/static/components/gameSearch.js]: Woo!
